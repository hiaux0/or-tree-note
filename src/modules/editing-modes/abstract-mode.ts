import {
  getComputedValueFromPixelString,
  getCssVar,
  getValueFromPixelString,
} from "../css/css-variables";

const CARET_NORMAL_CLASS = "caret-normal";
const CARET_INSERT_CLASS = "caret-insert";

export abstract class AbstractMode {
  children: NodeListOf<Element>;

  caretWidth: number;
  caretHeight: number;

  currentLineNumber: number = 0;
  currentCaretCol: number = 0;

  constructor(
    public parentElement: HTMLElement,
    public childSelector: string,
    public caretElement: HTMLElement
  ) {
    this.children = parentElement.querySelectorAll(this.childSelector);

    this.caretWidth = getCssVar("--caret-size-width");
    this.caretHeight = getCssVar("--caret-size-height");
  }

  cursorUp() {
    this.commenKeyFunctionality();
  }
  cursorDown() {
    this.commenKeyFunctionality();
  }
  cursorRight() {
    this.commenKeyFunctionality();

    const currentCaretLeft = getValueFromPixelString(
      this.caretElement.style.left
    );

    const newLeft = currentCaretLeft + this.caretWidth;
    const parentWidth = getValueFromPixelString(
      getComputedStyle(this.parentElement).width
    );

    /**
     * TODO: Only until word end
     */
    if (newLeft > parentWidth) {
      return;
    }

    this.caretElement.style.left = `${newLeft}px`;
  }
  cursorLeft() {
    this.commenKeyFunctionality();
  }

  commenKeyFunctionality() {
    this.resetCaretBlinking();
  }

  resetCaretBlinking() {
    this.caretElement.classList.remove("caret-blinking");
    /**
     * Needed to restart the animation
     * https://css-tricks.com/restart-css-animation/
     */
    void this.caretElement.offsetWidth;
    this.caretElement.classList.add("caret-blinking");
  }

  /** CARET */
  getCurrentCaretCol() {
    const curCarW = getComputedValueFromPixelString(this.caretElement, "width");
    /** -1 Caret style logic has always a start width -> Eg 10 / 10 = 1 */
    return curCarW / this.caretWidth - 1;
  }
}
