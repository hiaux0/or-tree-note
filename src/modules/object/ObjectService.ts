export class ObjectService {
  public static pick<T, Prop extends keyof T>(obj: T, keys: Prop[]) {
    const result: Partial<T> = {};
    keys.forEach((key) => {
      result[key] = obj[key];
    });

    return result;
  }
}
