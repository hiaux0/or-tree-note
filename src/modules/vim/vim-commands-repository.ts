import { ModifiersType, SPACE } from '../../resources/keybindings/app-keys';

export enum VIM_COMMAND {
  'backspace' = 'backspace',
  'clearLine' = 'clearLine',
  'createNewLine' = 'createNewLine',
  'cursorRight' = 'cursorRight',
  'cursorUp' = 'cursorUp',
  'cursorLeft' = 'cursorLeft',
  'cursorDown' = 'cursorDown',
  'cursorWordForwardEnd' = 'cursorWordForwardEnd',
  'cursorWordForwardStart' = 'cursorWordForwardStart',
  'cursorBackwordsStartWord' = 'cursorBackwordsStartWord',
  'cursorLineEnd' = 'cursorLineEnd',
  'cursorLineStart' = 'cursorLineStart',
  'delete' = 'delete',
  'deleteInnerWord' = 'deleteInnerWord',
  'deleteLine' = 'deleteLine',
  'enterInsertMode' = 'enterInsertMode',
  'enterNormalMode' = 'enterNormalMode',
  'enterVisualMode' = 'enterVisualMode',
  'indentRight' = 'indentRight',
  'indentLeft' = 'indentLeft',
  'joinLine' = 'joinLine',
  'newLine' = 'newLine',
  'replace' = 'replace',
  'space' = 'space',
  'toCharacterAtBack' = 'toCharacterAtBack',
  'toCharacterAt' = 'toCharacterAt',
  'toCharacterAfterBack' = 'toCharacterAfterBack',
  'toCharacterBefore' = 'toCharacterBefore',
  'type' = 'type',
  'yank' = 'yank',
  'visualDelete' = 'visualDelete',
  'visualInnerWord' = 'visualInnerWord',
  'visualStartLineWise' = 'visualStartLineWise',
  'visualMoveToOtherEndOfMarkedArea' = 'visualMoveToOtherEndOfMarkedArea',
}

export const VIM_MODE_COMMANDS = [
  VIM_COMMAND['enterInsertMode'],
  VIM_COMMAND['enterNormalMode'],
  VIM_COMMAND['enterVisualMode'],
];

export const VIM_COMMANDS = [
  VIM_COMMAND.newLine,
  VIM_COMMAND['backspace'],
  VIM_COMMAND['createNewLine'],
  VIM_COMMAND['cursorRight'],
  VIM_COMMAND['cursorUp'],
  VIM_COMMAND['cursorLeft'],
  VIM_COMMAND['cursorDown'],
  VIM_COMMAND['cursorWordForwardEnd'],
  VIM_COMMAND['cursorWordForwardStart'],
  VIM_COMMAND['cursorBackwordsStartWord'],
  VIM_COMMAND['cursorLineEnd'],
  VIM_COMMAND['cursorLineStart'],
  VIM_COMMAND['delete'],
  VIM_COMMAND['deleteLine'],
  VIM_COMMAND['deleteInnerWord'],

  VIM_COMMAND['indentRight'],
  VIM_COMMAND['indentLeft'],
  VIM_COMMAND['joinLine'],
  VIM_COMMAND['newLine'],
  VIM_COMMAND['replace'],
  VIM_COMMAND['space'],
  VIM_COMMAND['toCharacterAtBack'],
  VIM_COMMAND['toCharacterAt'],
  VIM_COMMAND['toCharacterAfterBack'],
  VIM_COMMAND['toCharacterBefore'],
  VIM_COMMAND['type'],
  VIM_COMMAND['yank'],
  // visual
  VIM_COMMAND['visualDelete'],
  VIM_COMMAND['visualInnerWord'],
  VIM_COMMAND['visualStartLineWise'],
  VIM_COMMAND['visualMoveToOtherEndOfMarkedArea'],
  // ...VIM_MODE_COMMANDS,
  VIM_COMMAND['enterInsertMode'],
  VIM_COMMAND['enterNormalMode'],
  VIM_COMMAND['enterVisualMode'],
];
export type VimCommandNames = keyof typeof VIM_COMMAND;

export const VIM_COMMANDS_THAT_CHANGE_TO_NORMAL_MODE = ['visualDelete'];

export interface VimCommand {
  key: string;
  command: VimCommandNames;
}

export interface SynonymKey {
  [key: string]: ModifiersType;
}
