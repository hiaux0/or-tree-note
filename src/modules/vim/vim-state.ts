import {
  Cursor,
  EMPTY_VIM_LINE,
  FoldMap,
  VimLine,
  VimMode,
  VimStateV2,
} from './vim-types';

export class VimStateClass {
  cursor: Cursor;
  foldMap: FoldMap;
  lines: VimLine[];
  mode: VimMode;
  visualStartCursor: Cursor;
  visualEndCursor: Cursor;
  deletedLinesIndeces: number[];
  commandName: string;

  constructor(public vimState: VimStateV2) {
    this.cursor = vimState.cursor;
    this.foldMap = vimState.foldMap;
    this.lines = vimState.lines;
    this.mode = vimState.mode ?? VimMode.NORMAL;
    this.visualStartCursor = vimState.visualStartCursor;
    this.visualEndCursor = vimState.visualEndCursor;
    this.deletedLinesIndeces = vimState.deletedLinesIndeces;
    this.commandName = vimState.commandName;
  }

  public static create(cursor: Cursor, lines?: VimLine[]) {
    return new VimStateClass({ cursor, lines });
  }

  public serialize(): VimStateV2 {
    return {
      cursor: this.cursor,
      foldMap: this.foldMap,
      lines: this.lines,
      mode: this.mode,
      visualStartCursor: this.visualStartCursor,
      visualEndCursor: this.visualEndCursor,
      deletedLinesIndeces: this.deletedLinesIndeces,
      commandName: this.commandName,
    };
  }

  public getLineAt(lineIndex: number) {
    const line = this.lines[lineIndex];
    return line;
  }

  public getActiveLine() {
    const active = this.getLineAt(this.cursor.line) ?? EMPTY_VIM_LINE;
    // const active = this.getLineAt(this.cursor.line);
    return active;
  }

  public getPreviousLine() {
    const previous = this.getLineAt(this.cursor.line - 1);
    return previous;
  }

  public updateLine(lineIndex: number, updated: string) {
    this.lines[lineIndex] = { text: updated };
  }

  public updateActiveLine(updated: string) {
    this.updateLine(this.cursor.line, updated);
  }
}
