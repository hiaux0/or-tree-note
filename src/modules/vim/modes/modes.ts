import { Logger } from "modules/debug/logger";
import { Cursor, VimCommandOutput, VimMode } from "../vim";

const logger = new Logger({ scope: "AbstractMode" });

export function isValidHorizontalPosition(
  cursorCol: number,
  activeInput: string
) {
  let result = true;
  if (cursorCol > activeInput.length - 1) {
    result = false;
  } else if (cursorCol < 0) {
    result = false;
  }
  return result;
}

export abstract class AbstractMode {
  /**
   * Firstly: The active input, is the eg. "active line".
   *
   * TODO:
   * - multiple cursors
   */
  activeInput: string;
  currentMode: VimMode;

  constructor(public vimCommandOutput: VimCommandOutput) {}

  executeCommand(
    commandName: string,
    commandValue: string,
    currentMode: VimMode
  ): VimCommandOutput {
    if (!this[commandName]) {
      logger.debug(
        [
          "No command '%s' found in %s Mode. ((modes.ts-executeCommand))",
          commandName,
          currentMode,
          this[commandName],
        ],
        { isError: true }
      );
    }

    const result = this[commandName](commandValue) as VimCommandOutput;

    if (result.text) {
      this.vimCommandOutput.text = result.text;
    }

    return result;
  }

  cursorRight(): VimCommandOutput {
    const updaterCursorCol = this.vimCommandOutput.cursor.col + 1;

    if (
      !isValidHorizontalPosition(updaterCursorCol, this.vimCommandOutput.text)
    ) {
      return {
        cursor: { ...this.vimCommandOutput.cursor },
        text: this.vimCommandOutput.text,
      };
    }

    this.vimCommandOutput.cursor.col = updaterCursorCol;
    return {
      cursor: { ...this.vimCommandOutput.cursor },
      text: this.vimCommandOutput.text,
    };
  }
  cursorLeft(): VimCommandOutput {
    const updaterCursorCol = this.vimCommandOutput.cursor.col - 1;

    if (
      !isValidHorizontalPosition(updaterCursorCol, this.vimCommandOutput.text)
    ) {
      return {
        cursor: { ...this.vimCommandOutput.cursor },
        text: this.vimCommandOutput.text,
      };
    }

    this.vimCommandOutput.cursor.col = updaterCursorCol;
    return {
      cursor: { ...this.vimCommandOutput.cursor },
      text: this.vimCommandOutput.text,
    };
  }
  cursorUp(): VimCommandOutput {
    this.vimCommandOutput.cursor.line -= 1;
    return {
      cursor: { ...this.vimCommandOutput.cursor },
      text: this.vimCommandOutput.text,
    };
  }
  cursorDown(): VimCommandOutput {
    this.vimCommandOutput.cursor.line += 1;
    return {
      cursor: { ...this.vimCommandOutput.cursor },
      text: this.vimCommandOutput.text,
    };
  }
}
