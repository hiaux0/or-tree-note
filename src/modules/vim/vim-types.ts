import { getRandomId } from 'common/random';

import { VimCommand, SynonymKey, VIM_COMMAND } from './vim-commands-repository';
import { VimStateClass } from './vim-state';

export interface KeyBindingModes {
  insert?: VimCommand[];
  normal: VimCommand[];
  visual?: VimCommand[];
  synonyms: SynonymKey;
}

export enum VimExecutingMode {
  'INDIVIDUAL' = 'INDIVIDUAL',
  'BATCH' = 'BATCH',
}

/**
 * Given a string 'Hello, World'
 * And I'm in normal mode
 * When I type "l"
 * Then the cursor should move one right
 */
export interface FindPotentialCommandReturn {
  targetCommand: VimCommand;
  potentialCommands: VimCommand[];
}

export type IndentationLevel = number;
export type Text = string;

export interface IndentationNode {
  text?: Text;
  indentation?: IndentationLevel;
}

export interface VimLine extends IndentationNode {
  text: Text;
  id?: string; // Migration_1
  // cursor?: Cursor;
}

export const EMPTY_VIM_LINE: VimLine = { text: '', id: getRandomId() };

export type FoldMap = Record<string, boolean>;

export type VimStateV2 = {
  cursor?: Cursor;
  lines?: VimLine[];
  mode?: VimMode;
  foldMap?: FoldMap;
  visualStartCursor?: Cursor;
  visualEndCursor?: Cursor;
  deletedLinesIndeces?: number[];
  commandName?: string;
};

export interface QueueInputReturn {
  vimState: VimStateClass | null;
  targetCommand: VIM_COMMAND;
  lines: VimLine[];
}

export interface Cursor {
  col: number;
  line: number;
}
export enum VimMode {
  'NORMAL' = 'NORMAL',
  'INSERT' = 'INSERT',
  'VISUAL' = 'VISUAL',
  'VISUALLINE' = 'VISUALLINE',
}
export type VimModeKeys = keyof typeof VimMode;

export interface VimOptions {
  keyBindings?: KeyBindingModes;
  leader?: string;
  vimPlugins?: VimPlugin[];
  indentSize?: number;
}

export interface VimPlugin {
  commandName: string;
  execute: (
    vimState?: VimStateClass,
    commandValue?: string
  ) => VimStateClass | void;
}
