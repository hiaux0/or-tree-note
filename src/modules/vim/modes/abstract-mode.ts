import { cloneDeep } from 'lodash';
import { Logger } from 'modules/debug/logger';
import {
  getFirstNonWhiteSpaceCharIndex,
  replaceAt,
  StringUtil,
} from 'modules/string/string';

import { VimStateClass } from '../vim-state';
import { VimState, VimMode, VimOptions, VimPlugin } from '../vim-types';

const logger = new Logger({ scope: 'AbstractMode' });

export interface TokenizedString {
  string: string;
  start: number;
  end: number;
  index: number;
}

export abstract class AbstractMode {
  /**
   * Firstly: The active input, is the eg. "active line".
   *
   * TODO:
   * - multiple cursors
   */
  currentMode: VimMode;

  constructor(
    public vimState: VimStateClass,
    public vimOptions: VimOptions = {}
  ) {
    // this.vimState.lines = this.vimState.lines;
  }

  executeCommand(
    commandName: string,
    commandInput: string,
    currentMode: VimMode
  ): VimStateClass {
    const targetVimPluginCommand = this.vimOptions.vimPlugins?.find(
      (plugin) => {
        return plugin.commandName === commandName;
      }
    );
    if (targetVimPluginCommand) {
      const pluginResult = this.executeVimPluginCommand(
        targetVimPluginCommand,
        commandInput
      );
      if (pluginResult) {
        return pluginResult;
      }
      return;
    }

    if (this[commandName] !== undefined) {
      logger.debug(
        [
          'No command "%s" found in %s Mode. ((modes.ts-executeCommand))',
          commandName,
          currentMode,
        ],
        { isError: true }
      );
    }

    const previousOutput = cloneDeep(this.vimState); // side effect, thus clone before executing command
    const result = this[commandName](commandInput) as VimStateClass;

    try {
      this.validateHorizontalCursor(result);
    } catch {
      logger.debug([
        'Not valid state. Returning to previous state: %o',
        previousOutput,
      ]);
      return previousOutput;
    }

    this.vimState = result;

    return result;
  }
  executeVimPluginCommand(targetVimPlugin: VimPlugin, commandValue: string) {
    return targetVimPlugin.execute(this.vimState, commandValue);
  }

  reTokenizeInput(input: string) {
    if (input === '') return;

    const tokenizedInput = tokenizeInput(input);

    logger.debug(['reTokenizeInput: %o', tokenizedInput], {
      onlyVerbose: true,
    });

    return tokenizedInput;
  }
  validateHorizontalCursor(vimState: VimStateClass) {
    const curCol = vimState.cursor.col + 1;
    const isValid = isValidHorizontalPosition(curCol, vimState.getActiveLine());

    if (!isValid) {
      try {
        logger.debug(
          [
            '[INVALID] Cursor col will be: %d, but should be between [0,%d].',
            curCol,
            vimState.getActiveLine().length,
          ],
          {
            isError: true,
          }
        );
      } catch {
        throw new Error();
      }
    }

    return isValid;
  }
  validateVerticalCursor(vimState: VimState) {
    const line = vimState.cursor.line + 1;
    const isValid = isValidVerticalPosition(line, this.vimState.lines);

    if (!isValid) {
      logger.debug(
        [
          '[INVALID] Line will be: %d, but should be between [0,%d].',
          line,
          this.vimState.lines.length,
        ],
        {
          isError: true,
        }
      );
      throw new Error();
    }

    return isValid;
  }

  /** ****** */
  /* Cursor */
  /** ****** */
  cursorRight(): VimStateClass {
    const updaterCursorCol = this.vimState.cursor.col + 1;

    if (
      !isValidHorizontalPosition(
        updaterCursorCol + 1,
        this.vimState.getActiveLine()
      )
    ) {
      return this.vimState;
    }

    this.vimState.cursor.col = updaterCursorCol;
    return this.vimState;
  }
  cursorLeft(): VimStateClass {
    const updaterCursorCol = this.vimState.cursor.col - 1;

    if (
      !isValidHorizontalPosition(
        updaterCursorCol + 1,
        this.vimState.getActiveLine()
      )
    ) {
      return this.vimState;
    }

    this.vimState.cursor.col = updaterCursorCol;
    return this.vimState;
  }
  cursorUp(): VimStateClass {
    const newCurLine = this.vimState.cursor.line - 1;
    const isValidVertical = isValidVerticalPosition(
      newCurLine + 1,
      this.vimState.lines
    );
    this.vimState.lines; /* ? */

    if (!isValidVertical) {
      return this.vimState;
    }

    const newActiveLine = this.vimState.lines[newCurLine];
    const isValidHorizontalAfterMovedVertically = isValidHorizontalPosition(
      this.vimState.cursor.col + 1,
      newActiveLine
    );

    if (!isValidHorizontalAfterMovedVertically) {
      // TODO: Call "$" to put cursor to end of line
      this.vimState.cursor.col = Math.max(newActiveLine.length - 1, 0);
    }

    const newActiveText = this.vimState.lines[newCurLine];

    // this.vimState.updateActiveLine(newActiveText);
    this.vimState.cursor.line = newCurLine;
    this.reTokenizeInput(newActiveText);

    return this.vimState;
  }
  cursorDown(): VimStateClass {
    const newCurLine = this.vimState.cursor.line + 1;
    const isValidVertical = isValidVerticalPosition(
      newCurLine + 1,
      this.vimState.lines
    );

    if (!isValidVertical) {
      return this.vimState;
    }

    const newActiveLine = this.vimState.lines[newCurLine];
    const isValidHorizontalAfterMovedVertically = isValidHorizontalPosition(
      this.vimState.cursor.col + 1,
      newActiveLine
    );

    if (!isValidHorizontalAfterMovedVertically) {
      // TODO: Call "$" to put cursor to end of line
      this.vimState.cursor.col = Math.max(newActiveLine.length - 1, 0);
    }

    const newActiveText = this.vimState.lines[newCurLine];

    this.vimState.cursor.line = newCurLine;
    this.reTokenizeInput(newActiveText);

    return this.vimState;
  }
  cursorWordForwardEnd(): VimStateClass {
    const tokenUnderCursor = this.getTokenUnderCursor();

    const isAtEnd = tokenUnderCursor?.end === this.vimState.cursor.col;
    const isNotAtEnd = tokenUnderCursor === undefined;

    let resultCol;
    if (isAtEnd) {
      const nextToken = this.getTokenAtIndex(tokenUnderCursor.index + 1);
      resultCol = nextToken.end;
    } else if (isNotAtEnd) {
      const nextToken = this.getNexToken();
      resultCol = nextToken.end;
    } else {
      resultCol = tokenUnderCursor.end;
    }

    if (resultCol) {
      this.vimState.cursor.col = resultCol;
    }

    return this.vimState;
  }
  cursorWordForwardStart(): VimStateClass {
    const nextToken = this.getNexToken();

    const isAtEnd = nextToken?.end === this.vimState.cursor.col;
    const isNotAtEnd = nextToken === undefined;

    let resultCol;
    if (isAtEnd) {
      const nextNextToken = this.getTokenAtIndex(nextToken.index + 1);
      resultCol = nextNextToken.end;
    } else if (isNotAtEnd) {
      const nextToken = this.getNexToken();
      resultCol = nextToken.end;
    } else {
      resultCol = nextToken.start;
    }

    if (resultCol) {
      this.vimState.cursor.col = resultCol;
    }

    return this.vimState;
  }
  cursorBackwordsStartWord(): VimStateClass {
    const tokenUnderCursor = this.getTokenUnderCursor(); /* ? */

    const isAtStart =
      tokenUnderCursor?.start === this.vimState.cursor.col; /* ? */
    const tokenNotUnderCursor = tokenUnderCursor === undefined;

    let resultCol;
    if (isAtStart) {
      const previousToken = this.getTokenAtIndex(tokenUnderCursor.index - 1);
      resultCol = previousToken.start;
    } else if (tokenNotUnderCursor) {
      const nextToken = this.getPreviousToken();
      resultCol = nextToken.start;
    } else {
      resultCol = tokenUnderCursor.start;
    }

    if (resultCol !== undefined) {
      this.vimState.cursor.col = resultCol;
    }

    return this.vimState;
  }
  cursorLineEnd(): VimStateClass {
    this.vimState.cursor.col = this.vimState.getActiveLine().length - 1;
    return this.vimState;
  }
  cursorLineStart(): VimStateClass {
    const nonWhiteSpaceIndex = getFirstNonWhiteSpaceCharIndex(
      this.vimState.getActiveLine()
    );

    this.vimState.cursor.col = nonWhiteSpaceIndex;
    return this.vimState;
  }
  toCharacterAtBack(commandInput: string): VimStateClass {
    const { cursor } = this.vimState;
    const text = this.vimState.getActiveLine();
    const currentTextFromStartToColumn = text.substring(0, cursor.col);
    const targetCharacterIndex = StringUtil.indexOfBack(
      currentTextFromStartToColumn,
      commandInput
    );

    if (targetCharacterIndex > -1) {
      this.vimState.cursor.col = targetCharacterIndex;
    }

    return this.vimState;
  }
  toCharacterAt(commandInput: string): VimStateClass {
    const { cursor } = this.vimState;
    const text = this.vimState.getActiveLine();
    const currentTextToEnd = text.substring(cursor.col);
    const targetCharacterIndex = currentTextToEnd.indexOf(commandInput);

    if (targetCharacterIndex > -1) {
      const finalNewIndex = cursor.col + targetCharacterIndex; // before substring + target index
      this.vimState.cursor.col = finalNewIndex;
    }

    return this.vimState;
  }
  toCharacterAfterBack(commandInput: string): VimStateClass {
    const { cursor } = this.vimState;
    const text = this.vimState.getActiveLine();
    const currentTextFromStartToColumn = text.substring(0, cursor.col);
    const targetCharacterIndex = StringUtil.indexOfBack(
      currentTextFromStartToColumn,
      commandInput
    );

    if (targetCharacterIndex > -1) {
      this.vimState.cursor.col = targetCharacterIndex + 1; // + 1 after char
    }

    return this.vimState;
  }
  toCharacterBefore(commandInput: string): VimStateClass {
    const { cursor } = this.vimState;
    const text = this.vimState.getActiveLine();
    const currentTextToEnd = text.substring(cursor.col);
    const targetCharacterIndex = currentTextToEnd.indexOf(commandInput);
    if (targetCharacterIndex > 0) {
      // ^ -1: stay, 0: stay, cos at beginning
      const finalNewIndex = cursor.col + targetCharacterIndex - 1; // before substring + target index - before character
      this.vimState.cursor.col = finalNewIndex;
    }

    return this.vimState;
  }

  /** ************** */
  /* Cursor Helpers */
  /** ************** */
  getTokenUnderCursor(): TokenizedString | undefined {
    const tokenizedInput = this.reTokenizeInput(this.vimState.getActiveLine());
    const targetToken = tokenizedInput.find((input) => {
      const curCol = this.vimState.cursor.col;
      const isUnderCursor = input.start <= curCol && curCol <= input.end;

      return isUnderCursor;
    });

    /* prettier-ignore */ logger.debug(['Token under curor: %o', targetToken], { onlyVerbose: true, });

    return targetToken;
  }
  getTokenAtIndex(index: number) {
    const tokenizedInput = this.reTokenizeInput(this.vimState.getActiveLine());

    if (index < 0) {
      index = 0;
    } else if (index > tokenizedInput.length - 1) {
      index = tokenizedInput.length - 1;
    }

    const targetToken = tokenizedInput[index];

    return targetToken;
  }
  getNexToken() {
    const tokenizedInput = this.reTokenizeInput(this.vimState.getActiveLine());
    const currentToken = this.getTokenUnderCursor();

    let nextIndex = NaN;
    if (currentToken == null) {
      nextIndex = this.getNextTokenFromCusor(tokenizedInput).index;
    } else {
      nextIndex = currentToken.index + 1;
    }

    const nextToken = tokenizedInput[nextIndex];

    if (nextToken == null) {
      return currentToken;
    }
    return nextToken;
  }
  private getNextTokenFromCusor(tokenizedInput: TokenizedString[]) {
    const currentCol = this.vimState.cursor.col;
    const maybeNext = tokenizedInput.find((input) => input.end >= currentCol);

    if (maybeNext == null) {
      return tokenizedInput[0];
    }

    return maybeNext;
  }
  getPreviousToken() {
    const tokenizedInput = this.reTokenizeInput(this.vimState.getActiveLine());
    const currentToken = this.getTokenUnderCursor();
    const previousToken = tokenizedInput[currentToken.index - 1];

    if (previousToken === undefined) {
      return currentToken;
    }
    return previousToken;
  }

  /** **** */
  /* Text */
  /** **** */
  indentRight(): VimStateClass {
    const { indentSize } = this.vimOptions;
    const spaces = ' '.repeat(indentSize);
    const updatedInput = `${spaces}${this.vimState.getActiveLine()}`;

    this.vimState.updateActiveLine(updatedInput);
    this.vimState.cursor.col += indentSize;
    this.vimState.lines[this.vimState.cursor.line] = updatedInput;
    this.reTokenizeInput(updatedInput);

    return this.vimState;
  }
  indentLeft(): VimStateClass {
    const { indentSize } = this.vimOptions;
    const text = this.vimState.getActiveLine();

    const stagedSubText = text.substring(0, indentSize);
    const whiteSpaceAtStartIndex = /\w/g.exec(stagedSubText);
    let numOfWhiteSpaceAtStart = stagedSubText.length;
    if (whiteSpaceAtStartIndex !== null) {
      numOfWhiteSpaceAtStart = whiteSpaceAtStartIndex.index;
    }

    const updatedInput = text.substring(numOfWhiteSpaceAtStart);
    this.vimState.updateActiveLine(updatedInput);
    this.vimState.cursor.col -= numOfWhiteSpaceAtStart;
    this.vimState.lines[this.vimState.cursor.line] = updatedInput;
    this.reTokenizeInput(updatedInput);

    return this.vimState;
  }
  delete(): VimStateClass {
    const updatedInput = replaceAt(
      this.vimState.getActiveLine(),
      this.vimState.cursor.col,
      ''
    );

    this.vimState.updateActiveLine(updatedInput);

    return this.vimState;
  }
  replace(commandInput: string): VimStateClass {
    const text = this.vimState.getActiveLine();
    const { col } = this.vimState.cursor;
    const replaced = replaceAt(text, col, commandInput);
    this.vimState.updateActiveLine(replaced);
    return this.vimState;
  }
  backspace(): VimStateClass {
    return this.vimState;
  }

  /** `o` */
  createNewLine(): VimStateClass {
    // add new line below
    const currentLine = this.vimState.getActiveLine();
    const newLineIndex = this.vimState.cursor.line + 1;
    const tempLines = [...this.vimState.lines];
    const numOfWs = StringUtil.getLeadingWhitespaceNum(currentLine);
    tempLines.splice(newLineIndex, 0, ' '.repeat(numOfWs));
    this.vimState.cursor.col = numOfWs;
    // put cursor below
    this.vimState.cursor.line = this.vimState.cursor.line + 1;
    this.vimState.lines = tempLines;
    // this.vimState.mode = VimMode.INSERT;

    /* prettier-ignore */ console.log('>>>> _ >>>> ~ file: abstract-mode.ts ~ line 502 ~ createNewLine');
    return this.vimState;
  }

  /** Lines */
  deleteLine(): VimStateClass {
    const curLine = this.vimState.cursor.line;
    // /* prettier-ignore */ console.trace('>>>> _ >>>> ~ file: abstract-mode.ts ~ line 504 ~ curLine', curLine);
    this.vimState.lines.splice(curLine, 1);

    let newCol = 0;
    if (this.vimState.getPreviousLine()) {
      newCol = Math.max(0, this.vimState.getPreviousLine().length - 1);
    } else {
      newCol = 0;
    }
    this.vimState.cursor.col = newCol;

    this.vimState.cursor.line = Math.max(curLine - 1, 0);
    const activeLine = this.vimState.getActiveLine();
    this.vimState.updateActiveLine(activeLine ?? '');

    //
    this.vimState.deletedLinesIndeces = [curLine];

    return this.vimState;
  }

  joinLine(): VimStateClass {
    const prev = this.vimState.getPreviousLine();
    const active = this.vimState.getActiveLine();
    const joined = prev.concat(active);

    const prevCursor = this.vimState.cursor.line - 1;
    this.vimState.updateLine(prevCursor, joined);
    this.deleteLine();
    this.vimState.cursor.col = prev.length;

    return this.vimState;
  }

  nothing(): VimStateClass {
    return this.vimState;
  }
}

export function isValidHorizontalPosition(
  cursorCol: number,
  activeInput: string
) {
  if (cursorCol === activeInput.length + 1) return true;

  const isBigger = cursorCol > activeInput.length;
  /**
   * Should be > technically, but conceptionally, cursor and text index are off by one.
   */
  const isZero = cursorCol === 0;
  const result = !isBigger && !isZero;

  return result;
}

export function isValidVerticalPosition(line: number, lines: string[]) {
  const isBigger = line > lines.length;
  /**
   * Should be > technically, but conceptionally, line and text index are off by one.
   */
  const isZero = line === 0;

  //
  const result = !isBigger && !isZero;
  return result;
}

export function tokenizeInput(input: string): TokenizedString[] {
  const regExp = /(\S+)/g;
  const matchResult: RegExpExecArray[] = [];
  let match: RegExpExecArray;

  while ((match = regExp.exec(input))) {
    matchResult.push(match);
  }

  const tokens = matchResult.map((result, resultIndex) => {
    const matchedString = result[0];
    const { index: matchIndex } = result;
    const token: TokenizedString = {
      string: matchedString,
      start: matchIndex,
      end: matchIndex + matchedString.length - 1,
      index: resultIndex,
    };
    return token;
  });

  logger.debug(['Tokens: %o', tokens], { onlyVerbose: true });

  return tokens;
}
