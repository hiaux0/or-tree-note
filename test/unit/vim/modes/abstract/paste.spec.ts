import { ObjectService } from '../../../../../src/modules/object/ObjectService';
import { AbstractMode } from '../../../../../src/modules/vim/modes/abstract-mode';
import { VIM_COMMAND } from '../../../../../src/modules/vim/vim-commands-repository';
import { VimStateClass } from '../../../../../src/modules/vim/vim-state';
import { VimMode } from '../../../../../src/modules/vim/vim-types';

const input = '012';
const cursor = { col: 1, line: 0 };

class TestingAbstractMode extends AbstractMode {}

describe('AbstractMode - paste', () => {
  const lines = input.split('\n').map((line) => ({ text: line }));
  let testingAbstractMode: AbstractMode;
  beforeEach(() => {
    const initialVimState = new VimStateClass({
      cursor,
      lines,
    });
    testingAbstractMode = new TestingAbstractMode(initialVimState);
  });

  it('1 line', async () => {
    const result = await testingAbstractMode.executeCommand(
      VIM_COMMAND['paste'],
      ['test'],
      VimMode.NORMAL
    );
    const expected = ObjectService.pick(result, ['lines', 'cursor']);
    expect(expected).toMatchSnapshot();
  });
  it('2 lines', async () => {
    const result = await testingAbstractMode.executeCommand(
      VIM_COMMAND['paste'],
      ['rs', 'st'],
      VimMode.NORMAL
    );
    const expected = ObjectService.pick(result, ['lines', 'cursor']);
    expect(expected).toMatchSnapshot();
  });
  it('3 lines', async () => {
    const result = await testingAbstractMode.executeCommand(
      VIM_COMMAND['paste'],
      ['rs', 'st', 'tu'],
      VimMode.NORMAL
    );
    const expected = ObjectService.pick(result, ['lines', 'cursor']);
    expect(expected).toMatchSnapshot();
  });
});
