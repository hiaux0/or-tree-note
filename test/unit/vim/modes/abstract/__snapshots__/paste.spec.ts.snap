// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractMode - paste 1 line 1`] = `
Object {
  "cursor": Object {
    "col": 1,
    "line": 0,
  },
  "lines": Array [
    Object {
      "text": "0test12",
    },
  ],
}
`;

exports[`AbstractMode - paste 2 lines 1`] = `
Object {
  "cursor": Object {
    "col": 1,
    "line": 0,
  },
  "lines": Array [
    Object {
      "text": "0rs",
    },
    Object {
      "text": "st12",
    },
  ],
}
`;

exports[`AbstractMode - paste 3 lines 1`] = `
Object {
  "cursor": Object {
    "col": 1,
    "line": 0,
  },
  "lines": Array [
    Object {
      "text": "0rs",
    },
    Object {
      "text": "st",
    },
    Object {
      "text": "tu12",
    },
  ],
}
`;
